lib/erl_type.ml

```diff
@@ -0,0 +1,73 @@
+open Base
+module Format = Caml.Format
+module Etf = Obeam.External_term_format
+module E = Etf_util
+open Common
+
+type qualifier = FloatQual | IntegerQual | NonemptyQual | PidQual | PortQual
+                 | ReferenceQual | UnknownQual
+                 (* | OtherQual of etf * etf *)
+[@@deriving show, sexp_of]
+
+let qualifier_of_etf etf =
+  let open Result in
+  let qual_of_atom = function
+    | "float" -> Ok FloatQual
+    | "integer" -> Ok IntegerQual
+    | "nonempty" -> Ok NonemptyQual
+    | "pid" -> Ok PidQual
+    | "port" -> Ok PortQual
+    | "reference" -> Ok ReferenceQual
+    | "unknown" -> Ok UnknownQual
+    | other -> Error (Failure(!%"qual_of_atom: unknown : %s" other))
+  in
+  E.atom_of_etf etf >>= fun atom ->
+  qual_of_atom atom
+
+type t_map_mandatoriness = Mandatory | Optional
+[@@deriving show, sexp_of]
+
+let mandatoriness_of_etf = function
+  | Etf.Atom "mand" -> Ok Mandatory
+  | Atom "opt" -> Ok Optional
+  | _ -> Error (Failure "mandatoriness_of_etf")
+
+(**
+{v
+-record(c, {tag			      :: tag(),
+	    elements  = []	      :: term(),
+	    qualifier = ?unknown_qual :: qual()}).
+
+-opaque erl_type() :: ?any | ?none | ?unit | #c{}.
+v}
+@see <https://github.com/erlang/otp/blob/OTP-21.1.1/lib/hipe/cerl/erl_types.erl>
+ *)
+type t =
+  | Any
+  | None
+  | Unit
+  | Atom of string list
```

Comment:
memo: `string list` seems strange for me, but in `erl_types.erl`, `atom` has a set
https://github.com/erlang/otp/blob/OTP-21.1.1/lib/hipe/cerl/erl_types.erl#L325